cmd/notary/integration_test.go

```diff
@@ -1526,3 +1527,317 @@ func TestWitness(t *testing.T) {
+	_, err = runCommand(t, tempDir, "-s", server.URL, "init", gunNoPublish)
+	require.NoError(t, err)
+	// list repo - exepect error
+	_, err = runCommand(t, tempDir, "-s", server.URL, "list", gunNoPublish)
+	require.NotNil(t, err)
+	require.Equal(t, err, nstorage.ErrMetaNotFound{Resource: "root"})
+}
+
+func TestClientTUFAddWithAutoPublish(t *testing.T) {
+	// -- setup --
+	setUp(t)
+
+	tempDir := tempDirWithConfig(t, "{}")
+	defer os.RemoveAll(tempDir)
+
+	server := setupServer()
+	defer server.Close()
+
+	tempFile, err := ioutil.TempFile("", "targetfile")
+	require.NoError(t, err)
+	tempFile.Close()
+	defer os.Remove(tempFile.Name())
+
+	var (
+		output          = ""
+		target          = "ShangXi"
+		targetNoPublish = "Shen-zinSu"
+		gun             = "MistsOfPandaria"
+	)
+	// -- tests --
+
+	// init repo with auto publish being enabled
+	_, err = runCommand(t, tempDir, "-s", server.URL, "init", "-p", gun)
+	require.NoError(t, err)
+	// add a target with auto publish being enabled
+	_, err = runCommand(t, tempDir, "-s", server.URL, "add", "-p", gun, target, tempFile.Name())
+	require.NoError(t, err)
+	// list repo - expect target
+	output, err = runCommand(t, tempDir, "-s", server.URL, "list", gun)
+	require.NoError(t, err)
+	require.True(t, strings.Contains(output, target))
+
+	// add a target without auto publish being enabled
+	//
+	// Use this test to guarantee that we won't break the normal add process.
+	_, err = runCommand(t, tempDir, "add", gun, targetNoPublish, tempFile.Name())
+	require.NoError(t, err)
+	// check status - expect the targetNoPublish
+	output, err = runCommand(t, tempDir, "status", gun)
+	require.NoError(t, err)
+	require.True(t, strings.Contains(output, targetNoPublish))
+	// list repo - expect only the target, not the targetNoPublish
+	output, err = runCommand(t, tempDir, "-s", server.URL, "list", gun)
+	require.NoError(t, err)
+	require.True(t, strings.Contains(output, target))
+	require.False(t, strings.Contains(output, targetNoPublish))
+}
+
+func TestClientTUFRemoveWithAutoPublish(t *testing.T) {
+	// -- setup --
+	setUp(t)
+
+	tempDir := tempDirWithConfig(t, "{}")
+	defer os.RemoveAll(tempDir)
+
+	server := setupServer()
+	defer server.Close()
+
+	tempFile, err := ioutil.TempFile("", "targetfile")
+	require.NoError(t, err)
+	tempFile.Close()
+	defer os.Remove(tempFile.Name())
+
+	var (
+		output              = ""
+		target              = "ShangXi"
+		targetWillBeRemoved = "Shen-zinSu"
+		gun                 = "MistsOfPandaria"
+	)
+	// -- tests --
+
+	// init repo with auto publish being enabled
+	_, err = runCommand(t, tempDir, "-s", server.URL, "init", "-p", gun)
+	require.NoError(t, err)
+	// add a target with auto publish being enabled
+	_, err = runCommand(t, tempDir, "add", "-s", server.URL, "-p", gun, target, tempFile.Name())
+	require.NoError(t, err)
+	_, err = runCommand(t, tempDir, "add", "-s", server.URL, "-p", gun, targetWillBeRemoved, tempFile.Name())
```

Comment:
Another test idea: try an autopublish command without giving a server URL, and check that we failed to publish.  And ensure that the change is still staged